### ISSUE TITLE
Reads of large twcs partitions scan many sstables in parallel

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#6418]
Consider this query for a TWCS table:

    SELECT * FROM tab
    WHERE pk = :pk AND ck > :lower

Currently, we will read from all sstables in all time windows, assuming we the partition exists for every time window. This is expensive - we'll read an index block at least for every sstables. But usually data will only come from a single sstable, since each ck range in confined to a window. Once in a while we'll need to read from two sstables (when crossing time windows), and sometimes we'll need to read sequentially from multiple sstables (a very sparse partition). But we don't need to read from all sstables at once, so the current solution has a large read amplification for the first read, and large memory amplification for following reads, assuming #1865 is effective (which is might not be, considering that there is large memory consumption).

Once #4042 is fixed, we will filter out sstables that have their maximum ck greater than or equal to `:lower`.  But if `:lower` is sufficiently low, then we still read from many sstables in parallel.

This can be fixed if we employ an idea similar to the incremental selector (bfc74f13129ae9f5c3e009745a2d05f088f82180). This can feed sstables to the reader in order of their [min, max] range. Note that while the original incremental selector used the token + partition key to sort sstables,  in order to improve partition range scans, the new selector sorts by clustering keys, to support single-partition scans.

**Comment by avikivity** [comment]
@slivne this is needed for large partition TWCS, esp. CDC

**Comment by denesb** [comment]
Refs: https://github.com/scylladb/scylla/issues/5124

**Comment by denesb** [comment]
A very similar idea was discussed and ultimately deemed as not worth doing in the above issue, although in a more general context. @avikivity do you intend to employ this optimization only for TWCS tables, similarly how the existing incremental selector is only used for LCS?

**Comment by avikivity** [comment]
Looks like this is a duplicate of #5124. Nothing changed, we just under-estimated the importance. TWCS will often see partitions that extend across all time windows.

This optimization can't help anything other than TWCS, because for other compaction strategies, the clustering keys will be evenly spread out among all sstables as compation mixes them up. In fact maybe we shouldn't even collect the statistics for those compaction strategies.

**Comment by avikivity** [comment]
@piodul note that this has a hard dependency on #4042. Implementation details can be found in #1935 commits and @denesb mind.

**Comment by kbr-** [comment]
I need some clarification.

This issue is not about the **set of sstables** being read --- there was such an issue, i.e. we were reading sstables unnecessarily when we could quickly rule out some of them, but it was fixed already, as described by https://github.com/scylladb/scylla/issues/4042.

Thus, the set of sstables is good enough. Our problem is with **how we read this set**, i.e. that we read them in parallel, while we should be reading them sequentially; is that correct? @avikivity 

Another question: in the example you've given, you specify a single partition key. But in the CDC use case (which is the main motivation in pursuing this issue AFAIU) we want to query multiple partitions at once, e.g.
```
select * from t where pk in (0, 1) and ck > 42;
```
now, within each partition, we would probably want to read the sstables for this partition sequentially. However, if we're reading multiple partitions, and the data resides in different sstables, we may want, or even need (?) to read the sstables in parallel, right?

**Comment by raphaelsc** [comment]
> I need some clarification.
> 
> This issue is not about the **set of sstables** being read --- there was such an issue, i.e. we were reading sstables unnecessarily when we could quickly rule out some of them, but it was fixed already, as described by #4042.
> 
> Thus, the set of sstables is good enough. Our problem is with **how we read this set**, i.e. that we read them in parallel, while we should be reading them sequentially; is that correct? @avikivity
> 
> Another question: in the example you've given, you specify a single partition key. But in the CDC use case (which is the main motivation in pursuing this issue AFAIU) we want to query multiple partitions at once, e.g.
> 
> ```
> select * from t where pk in (0, 1) and ck > 42;
> ```
> 
> now, within each partition, we would probably want to read the sstables for this partition sequentially. However, if we're reading multiple partitions, and the data resides in different sstables, we may want, or even need (?) to read the sstables in parallel, right?

reader advances one partition at a time, so we could still incrementally read a given partition by only looking at SSTables that overlap with the current clustering key being read (alike partitioned_sstable_set), which in TWCS will only be a single SSTable because the clustering key reflects the time of the data and TWCS works so that data in a given time window is stored in a single SSTable.

**Comment by denesb** [comment]
> This issue is not about the **set of sstables** being read --- there was such an issue, i.e. we were reading sstables unnecessarily when we could quickly rule out some of them, but it was fixed already, as described by #4042.
> 
> Thus, the set of sstables is good enough. Our problem is with **how we read this set**, i.e. that we read them in parallel, while we should be reading them sequentially; is that correct? @avikivity

Yes, we want to read the sstables incrementally that is at each clustering key we want to have only those sstables open whose clustering range intersects with said key.

> 
> Another question: in the example you've given, you specify a single partition key. But in the CDC use case (which is the main motivation in pursuing this issue AFAIU) we want to query multiple partitions at once, e.g.
> 
> ```
> select * from t where pk in (0, 1) and ck > 42;
> ```
> 
> now, within each partition, we would probably want to read the sstables for this partition sequentially. However, if we're reading multiple partitions, and the data resides in different sstables, we may want, or even need (?) to read the sstables in parallel, right?

A query like this would actually result in two different read request sent to replicas, each for just one partition. And anyway, like @raphaelsc said, the read advances one partition at a time.

**Comment by avikivity** [comment]
> I need some clarification.
> 
> This issue is not about the **set of sstables** being read --- there was such an issue, i.e. we were reading sstables unnecessarily when we could quickly rule out some of them, but it was fixed already, as described by #4042.


Yes.

> 
> Thus, the set of sstables is good enough. Our problem is with **how we read this set**, i.e. that we read them in parallel, while we should be reading them sequentially; is that correct? @avikivity


Yes. With the intent that we will only read one of them for the current page.


> 
> Another question: in the example you've given, you specify a single partition key. But in the CDC use case (which is the main motivation in pursuing this issue AFAIU) we want to query multiple partitions at once, e.g.
> 
> ```
> select * from t where pk in (0, 1) and ck > 42;
> ```
> 
> now, within each partition, we would probably want to read the sstables for this partition sequentially. However, if we're reading multiple partitions, and the data resides in different sstables, we may want, or even need (?) to read the sstables in parallel, right?

No, you need to read partition 0 and exhaust it, then 1 and exhaust it.

But I don't think the query is correct. You need to read from partition 0 where ck > last_p0_ck_seen, and from partition 1 where ck > last_p1_ck_seen, they won't be the same. So it's enough to focus on a single partition, and sort sstables by clustering key lower limit.

@denesb I think the incremental selector is only called at partition boundaries, is this correct?

**Comment by denesb** [comment]
> @denesb I think the incremental selector is only called at partition boundaries, is this correct?

Correct.

**Comment by kbr-** [comment]
Ok, thanks for the answers.

@raphaelsc 
> which in TWCS will only be a single SSTable because the clustering key reflects the time of the data and TWCS works so that data in a given time window is stored in a single SSTable.

The current time window may be stored in multiple sstables (they are compacted using STCS). I imagine that if there are a lot of writes and memtables are flushed frequently, and the writes sometimes come out of order (not in the order of timestamps), we may get an sstable containing write with timestamp `T1` flushed *later* than some other sstable containing write with timestamp `T2` where `T2 >= T1`; in particular, we may even get the same timestamp appearing in multiple sstables. And if we take an interval of timestamps, the chance that timestamps from this interval appear in multiple sstables rises.

@avikivity 
> But I don't think the query is correct. You need to read from partition 0 where ck > last_p0_ck_seen, and from partition 1 where ck > last_p1_ck_seen, they won't be the same. So it's enough to focus on a single partition, and sort sstables by clustering key lower limit.

Some time ago we've been discussing on CDC calls the possibility of a certain optimization in which we query the minimum of last_`pi`_ck_seen for multiple `pi`s, because it might be cheaper to just throw out some of the queried data instead of sending multiple requests. Of course this would only be done for `pi`s that have the same replica sets -- it assumes that we can reduce the number of round trips this way.

And maybe there are other reasons we want to make such "grouped queries". We even introduced a notion of grouping into stream ID format itself, for the purpose of grouped queries, if I understand correctly. @elcallio @haaawk could you confirm/deny?

However, @denesb said
> A query like this would actually result in two different read request sent to replicas, each for just one partition

are you sure? Even if they are in the same vnode? Sounds suboptimal.

**Comment by avikivity** [comment]
I don't think reducing query count is so important. In terms of disk operations it's the same since the partitions are stored separately. So you only reduce some coordinator-side processing, at the cost of increasing complexity.

**Comment by denesb** [comment]
Look at https://github.com/scylladb/scylla/blob/f76c519c1fc9f511383024b83f3b015833a65736/service/storage_proxy.cc#L3918 Don't bother grouping because storage proxy will throw your effort out of the window. Yes it is suboptimal in the sense that in certain combination of cluster topology and queried partitions some (parallel mind you) round-trips are saved, but IMO the complexity is not worth it.

**Comment by kbr-** [comment]
I imagine if there are, say, 100 stream IDs in a vnode, then having to send one message there-and-back instead of 100 (in the "normal" case; let's say that the data fits on a single page) could increase performance substantially.

**Comment by denesb** [comment]
Can't you use range-scan to read those 100 stream IDs? Or they are not guaranteed to form a continuous range?

**Comment by kbr-** [comment]
> in certain combination of cluster topology and queried partitions 

Ahh, but for CDC this combination would be the common case.
The number of stream IDs is around N = (number of vnodes) * (number of shards per node); for each vnode we have around (number of shards per node) stream IDs in this vnode. We want to query all of them periodically. Now we could perform N queries in each period, or we could perform (number of vnodes) grouped queries, where each of these grouped queries makes one round-trip.

**Comment by kbr-** [comment]
> Can't you use range-scan to read those 100 stream IDs? Or they are not guaranteed to form a continuous range?

so something like
```
where token(cdc$stream_id) > X and token(cdc$stream_id) < Y and cdc$time > T
```
? (`cdc$time` is the clustering key here)

**Comment by denesb** [comment]
Yes, this should work, you can supply a slice for range scans as well.

**Comment by haaawk** [comment]
This is weird @denesb because we observed a performance gain when we switched from querying each stream with separate query to querying groups of streams all in the same vnode.

**Comment by denesb** [comment]
> > in certain combination of cluster topology and queried partitions
> 
> Ahh, but for CDC this combination would be the common case.
> The number of stream IDs is around N = (number of vnodes) * (number of shards per node); for each vnode we have around (number of shards per node) stream IDs in this vnode. We want to query all of them periodically. Now we could perform N queries in each period, or we could perform (number of vnodes) grouped queries, where each of these grouped queries makes one round-trip.

So you have one stream ID per shard per vnode. In this case each partition would be read from a different shard anyway. So instead of combining the results on the coordinator node, you do this on the replica.

**Comment by denesb** [comment]
> This is weird @denesb because we observed a performance gain when we switched from querying each stream with separate query to querying groups of streams all in the same vnode.

It is probably cheaper to execute one CQL query than 100 of them, but on the replica level they are equivalent.

**Comment by kbr-** [comment]
> So you have one stream ID per shard per vnode. In this case each partition would be read from a different shard anyway. So instead of combining the results on the coordinator node, you do this on the replica.

Yes. But in the end, while disk usage seems to be the same, we may be putting much less strain on the network.

---

So all in all, there **is** a way to query multiple partitions, and at the same time, a range of clustering keys. Whether it's by grouped IN queries (dunno if this works, as Botond observed, `storage_proxy` may cancel our effors), or by token range queries (this should work).

About reading sstables sequentially vs in parallel:
since we're dealing with SCTS in the current window, not with LCS, it means that we don't have a run of sstables partitioned w.r.t token ranges; we're dealing with multiple sstables with overlapping tokens. So we cannot use the incremental selector (I was thinking if it could be somehow composed with this new strategy).

It sounds like we may have to open multiple sstables, because we want to read partitions in token order, and we don't know in which sstable the first partition's data is stored, and so on.

We may even need to do that for just a single-partition query, because the writes may arrive out of order, so we may have multiple sstables with overlapping timestamp ranges (in the current time window at least).

But still, for the common case, the number of sstables being read at the same time should be reduced.

**Comment by denesb** [comment]
We **can** and should use the incremental selector because if the read range spawns multiple windows we can prevent opening all sstables for all windows and just open those (or a subset of) a single window at a time. This can make a big difference.

**Comment by kbr-** [comment]
I was referring to the per-partition-incremental-selector which can be used only for LCS (as you said before).

**Comment by denesb** [comment]
It can be used with any CS which arranges sstables in a way that they are mostly disjoint. TWCS is one such CS.

**Comment by kbr-** [comment]
> disjoint

with respect to the ring position, right?
```
    // Used to incrementally select sstables from sstable set using ring-position.
    // sstable set must be alive during the lifetime of the selector.
    class incremental_selector {
        std::unique_ptr<incremental_selector_impl> _impl;
        dht::ring_position_comparator _cmp;
```

here we're dealing with sstables that are "disjoint", but not w.r.t ring position (because the current window uses SCTS); instead, they are (mostly) disjoint w.r.t clustering keys.

**Comment by denesb** [comment]
Right, actually TWCS doesn't necessarily have sstable disjoint on the partition level. This creates an interesting situation. Given the following dataset:
```
        |        TW1       |
sst1 P1 |c1 c2 c3 c4 c5 c6 | 
sst1 P2 |            c5 c6 |              
sst2 P2 |c1    c3          |       
sst2 P3 |   c2    c4 c5    | 
       -----------------------> clustering domain

```

If we are reading all three partitions, we will open readers for all two sstables when reading P1, because their clustering ranges overlap. However we will only read from one reader. At the same time we wouldn't be able to close the reader for sst2 because it has data for later parts of the range. This problem wouldn't exist if we were reading only one partition, then we can close any sstable reader that turns out to not have data for the current partition.

**Comment by avikivity** [comment]
Yes. Note that CDC *usually* reads from just one window (near the current time), but if some reader falls behind and tries to catch up, it will have to read many windows. So it's better to send one query per partition.

**Comment by kbr-** [comment]
I've finally found some time to actually sit on this issue, and I noticed the following problem.

We must open all of the readers, because we must first consume `partition_start` (which might contain a partition tombstone) and `static_row` (if exists) from each reader. Thus I deduced that an approach with a selector that creates new sstable readers when we're in the middle of a partition won't work (I also don't like this whole selector abstraction).

From what I understand, the problem we're trying to solve here is large memory consumption by all those sstable readers opened at the same time, due to large read-ahead performed by each reader.

Observation: after opening the reader, we don't have to read anything else except `partition_start` and `static_row` until, in the merging algorithm, we reach a position in partition that the reader may contain (which we decide based on the min/max ck metadata). 

Hence what I believe could work, is some kind of lazy reader abstraction that, when created, makes a sequence of smaller reads until it finds that either it already buffered the partition tombstone and static row (which hopefully are small) or the buffer already grew to the usual read-ahead value. After returning partition tombstone and static row, it will wait until the next `operator()` call and only then perform the usual read-ahead.

The merging algorithm would then open all readers, consume only `partition_start`+`static_row` from each, and then take care not to call `operator()` on readers that it knows to contain only positions greater than the current one. It could store these "dormant" readers in some kind of queue sorted by their min-ck and move them from the queue to the usual fragment heap when it decides the time is right.

BTW. `mutation_reader_merger` is extremely complex. I don't think that adding another bunch of `if`s to it is a good idea; also, since we're dealing with single-partition reads here, we could probably do fine with a simpler abstraction (e.g. `_reader_heap` is unnecessary). Hence I'm thinking about writing a new type of merger to solve our current problem. Maybe some parts of the logic could be refactored out into freestanding functions and then shared, but we will see.

**Comment by kbr-** [comment]
Another way would be... go with the selector approach, take the `partition_start` returned by the first queried reader, assume there are no partition tombstones nor static rows, and ignore `partition_start`s returned by other readers (when we open them in the middle of partition read).

This doesn't work for all sstables since the no-tombstone no-static-row assumption is not true in general.

But for CDC log tables it could work --- we don't have static columns there and we don't insert any partition tombstones. So we could only enable the optimization for CDC log tables. But we would have to remember not to change this assumption in the future. I guess the merging reader could make an `assert` at the right moment so we quickly catch if the assumption is false in unit tests.

**Comment by avikivity** [comment]
> I've finally found some time to actually sit on this issue, and I noticed the following problem.
> 
> We must open all of the readers, because we must first consume `partition_start` (which might contain a partition tombstone) and `static_row` (if exists) from each reader in order to merge them. 

TWCS never has static rows or partition tombstones. So it's enough to detect that there are no partition tombstones, and that there are no static rows.

There is sstable metadata that tells you the tombstone timestamps. From this you can deduce there are no partition tombstones (TWCS tables also have no other tombstones).

Lack of static rows can be seen in the schema.

**Comment by haaawk** [comment]
Maybe we could store in min ck metadata a special marker if sstable contains partition deletion and another for static row. This way we would know whether we need to open this sstable to get partition deletion and/or static row.

**Comment by haaawk** [comment]
@avikivity Is there anything preventing users from deleting the whole partition with CQL operation when using TWCS? That's something new for me.

**Comment by avikivity** [comment]
Nothing, they just don't do it. Standard practice with TWCS is to fully rely on TTL.

So we can detect that they're following standard practice, and enable the optimization only if they do.

**Comment by avikivity** [comment]
> 
> But for CDC log tables it could work --- we don't have static columns there and we don't insert any partition tombstones. So we could enable the optimization only for CDC log tables. But we would have to remember not to change this assumption in the future. I guess the merging reader could make an `assert` at the right moment so we quickly catch if the assumption is false in unit tests.


It's not just you, it's everyone. So we can just detect that the assumption is correct (looking at the schema and metadata) and use the new selector only if so.

**Comment by denesb** [comment]
> BTW. `mutation_reader_merger` is extremely complex. I don't think that adding another bunch of `if`s to it is a good idea; also, since we're dealing with single-partition reads here, we could probably do fine with a simpler abstraction (e.g. `_reader_heap` is unnecessary). Hence I'm thinking about writing a new type of merger to solve our current problem. Maybe some parts of the logic could be refactored out into freestanding functions and then shared, but we will see.

I also think we shouldn't hammer this into the existing merger. The combining reader was written with multiple fragment producers in mind, I think we should create a new one for this use-case.

**Comment by avikivity** [comment]
I agree.

**Comment by raphaelsc** [comment]
> Looks like this is a duplicate of #5124. Nothing changed, we just under-estimated the importance. TWCS will often see partitions that extend across all time windows.
> 

It's recommended that TWCS users avoid that though by aligning partitions to time windows (for example, if window is 1 day, put day_of_year into partition key), but I think many users just don't do that.

**Comment by avikivity** [comment]
This is a major optimization, so not backporting. Upgrade to enjoy.

**Comment by vladzcloudius** [comment]
> > Looks like this is a duplicate of #5124. Nothing changed, we just under-estimated the importance. TWCS will often see partitions that extend across all time windows.
> 
> It's recommended that TWCS users avoid that though by aligning partitions to time windows (for example, if window is 1 day, put day_of_year into partition key), but I think many users just don't do that.

@raphaelsc is right. The data model described above is going to be an anti-pattern for Scylla because this data model has a potential to generate huge partitions which in line would create CPU hot spots.

We always recommend to have time component as a part of a partition key like Raphael described above.

@kbr- Do we have any performance/benchmark/testing numbers that would allow us to quantify the benefit from this patch?

**Comment by kbr-** [comment]
> @kbr- Do we have any performance/benchmark/testing numbers that would allow us to quantify the benefit from this patch?

Manual experiment to check memory usage, described in the PR cover letter here: https://github.com/scylladb/scylla/pull/7437
Plus a microbenchmark that tests the underlying machinery - mutation reader merger for TWCS sstables - and compares it to the generic reader merger used in other compaction strategies, described in the PR cover letter here: https://github.com/scylladb/scylla/pull/7688
But I'm not aware of any "system benchmarks".

**Comment by vladzcloudius** [comment]
> > @kbr- Do we have any performance/benchmark/testing numbers that would allow us to quantify the benefit from this patch?
> 
> Manual experiment to check memory usage, described in the PR cover letter here: #7437 Plus a microbenchmark that tests the underlying machinery - mutation reader merger for TWCS sstables - and compares it to the generic reader merger used in other compaction strategies, described in the PR cover letter here: #7688 But I'm not aware of any "system benchmarks".

Thanks, Kamil.
So, as expected the patch improves the memory consumption by 460K when there 8 buckets. So, if the amount of buckets is 20 it would be probably something like 2MB less data read.
If we issue those reads concurrently (as we should) the added latency would probably not be measurable if there is spare I/O concurrency to do that. 

If the shard's I/O is going to be loaded to the point that all those 19 extra reads need to be issued serially then it may add ~10-20ms to the latency.

However I doubt that this use case is interesting because in that case (total I/O overload) you probably wouldn't care much about those extra 10-20ms since your latency are probably going to be at 100s of ms due to queuing.

I think it would be interesting to have a full system test demonstrating the improvement coming from this patch.

We should also check if this patch doesn't introduce a regression because not all use cases that use TWCS are going to read data for "a single or maybe 2 sstables" as this patch series assumes. If reads involve more sstables (e.g. the whole partition) then the fact that we are not going to open all sstables and use a regular merging reader can actually make the query to be slower.

@avikivity @kbr- WDYT?

**Comment by avikivity** [comment]
Most queries are going to page. It's safer to assume that (and maybe add latency)  Than assume the opposite (and throw away reads).

**Comment by vladzcloudius** [comment]
> Most queries are going to page. It's safer to assume that (and maybe add latency) Than assume the opposite (and throw away reads).

Not sure I understand how this is related. Wouldn't this patch get triggered for a paged query too, @avikivity? And why do you assume that a single page would be possible to serve from a single sstable? 

The assumption that TWCS is only used with a schema that has a write-time-based clustering order is incorrect.
The condition for TWCS are as follows:
* all data is TTLed (preferably the same)
* No updates
* No deletes

**Comment by kbr-** [comment]
> We should also check if this patch doesn't introduce a regression because not all use cases that use TWCS are going to read data for "a single or maybe 2 sstables" as this patch series assumes. 

That's not what the patch series assumes.
If the data comes from one or two sstables, the implemented optimization doesn't matter. Because we would already read only one or two sstables thanks to the partition key filters which would cause other sstables to not be opened at all.
The intended use case for the optimization is when the data lives in many sstables and we need to traverse them all.

**Comment by vladzcloudius** [comment]
> > We should also check if this patch doesn't introduce a regression because not all use cases that use TWCS are going to read data for "a single or maybe 2 sstables" as this patch series assumes.
> 
> That's not what the patch series assumes. If the data comes from one or two sstables, the implemented optimization doesn't matter. Because we would already read only one or two sstables thanks to the partition key filters which would cause other sstables to not be opened at all. The intended use case for the optimization is when the data lives in many sstables and we need to traverse them all.

This patch assumes that a query result doesn't need to be merged - and what if it does, @kbr- ?

**Comment by avikivity** [comment]
Why would it be inefficient? Less work would need to be done overall, since the sstables are opened in series and not merged (merging it O(log n) in the number of windows, though gallop mode reduces it).

It would take more time, but TWCS queries that read through many windows can't be latency sensitive.

**Comment by vladzcloudius** [comment]
> Why would it be inefficient? Less work would need to be done overall, since the sstables are opened in series and not merged (merging it O(log n) in the number of windows, though gallop mode reduces it).

If clustering is not write-time-based then merging requires reading from all relevant sstables. In the worst case it may be that you need to read each next row from another sstable. If that would require you to close the current sstable file descriptor and then open the FD for the sstable you need to read the next row from it will become very slow and inefficient.

> 
> It would take more time, but TWCS queries that read through many windows can't be latency sensitive.

We should not reason about it because it's up to a user to decide what's latency sensitive and what's not. And since TWCS exists for some time in Enterprise this new heuristics may be introduces a regression.
The easiest way is to have a way to disable this heuristic if needed.

**Comment by avikivity** [comment]
> > Why would it be inefficient? Less work would need to be done overall, since the sstables are opened in series and not merged (merging it O(log n) in the number of windows, though gallop mode reduces it).
> 
> If clustering is not write-time-based then merging requires reading from all relevant sstables. In the worst case it may be that you need to read each next row from another sstable. If that would require you to close the current sstable file descriptor and then open the FD for the sstable you need to read the next row from it will become very slow and inefficient.

We don't close and open file descriptors dynamically, only when the sstable is created and deleted. Even if we did, efficiency is still better (appending vs. merging), but latency is worse.

> 
> > It would take more time, but TWCS queries that read through many windows can't be latency sensitive.
> 
> We should not reason about it because it's up to a user to decide what's latency sensitive and what's not. And since TWCS exists for some time in our releases and this new heuristics may be introducing a regression. The easiest way is to have a way to disable this heuristic if needed.


For sure we can add more configuration parameters, but that just blows the testing matrix out of proportions.

**Comment by vladzcloudius** [comment]
> For sure we can add more configuration parameters, but that just blows the testing matrix out of proportions.

It's not about our testing matrix (it's actually irrelevant for this discussion) - it's about users that already use Scylla and can start seeing a performance degradation once they get Scylla with this new heuristics. And we definitely don't want that.

So, we can either prove that there is no performance degradation in every possible use case or we can introduce a parameter that allows disabling this heuristics (or better - enabling it).

But we can't do nothing and AFAIU we did none of the above.

**Comment by vladzcloudius** [comment]
BTW, how this heuritistics does when the data that is read is in the current time window which is managed by STCS, @kbr- ?

**Comment by kbr-** [comment]
> This patch assumes that a rows from different sstables don't need to be merged - and what if they do, @kbr- ?

No, it doesn't assume that. The `clustering_order_reader_merger` which is a huge part of this patch exists precisely to merge rows from different sstables. It just does it better than the "regular merger" when sstables have small clustering key overlaps. When they have big overlaps, it should perform similarly to the "regular merger".

**Comment by kbr-** [comment]
> BTW, how this heuritistics does when the data that is read is in the current time window which is managed by STCS, @kbr- ?

You mean "what does it do" or "how does it perform"? What does it do - it does the same thing. We read from all sstables that may contain the data we look for. The difference is that we don't open sstable readers until we need them, and we close them as soon as we no longer need them.
How does it perform - not sure, this needs bigger benchmarks.

**Comment by vladzcloudius** [comment]
>>    This patch assumes that a rows from different sstables don't need to be merged - and what if they do, @kbr- ?

> No, it doesn't assume that. The clustering_order_reader_merger which is a huge part of this patch exists precisely to merge rows from different sstables. It just does it better than the "regular merger" when sstables have small clustering key overlaps. When they have big overlaps, it should perform similarly to the "regular merger".

That's exactly what I meant then I said that "it assumes that a rows from different sstables don't need to be merged". The more precise wording would be: "it assumes that rows from different sstables don't need to be merged *often*".

> You mean "what does it do" or "how does it perform"? What does it do - it does the same thing. We read from all sstables that may contain the data we look for. The difference is that we don't open sstable readers until we need them, and we close them as soon as we no longer need them.
How does it perform - not sure, this needs bigger benchmarks.

I was asking about performance testing. 
And the fact that we neither have this testing nor have a way to disable this new feature concerns me a lot.

**Comment by vladzcloudius** [comment]
@slivne @eliransin FYI
